### PR TITLE
fix re-projection by introducing StagingPass

### PR DESCRIPTION
**Pull request by galister**
StagingPass is now used to copy the hardwareBuffer into a GL_TEXTURE_2D when FFR is disabled.

### CODE REVIEW

File: alvr/client_openxr/src/lib.rs [hunk 1]
```diff
@@ -636,9 +645,7 @@ pub fn entry_point() {
             swapchains[0].wait_image(xr::Duration::INFINITE).unwrap();
             swapchains[1].wait_image(xr::Duration::INFINITE).unwrap();
 
-            if !hardware_buffer.is_null() {
-                alvr_client_opengl::render_stream(hardware_buffer, [left_view_idx, right_view_idx]);
-            }
+            alvr_client_opengl::render_stream(hardware_buffer, [left_view_idx, right_view_idx]);
```
Comment (galister) [inline alvr/client_openxr/src/lib.rs:648]: not sure if we can have a case where `!hardware_buffer.is_null() && history_views.is_some()`
we probably want to force a re-projection in that case?
either way, I used it with this logic for ~1h yesterday and i did not see any flickers.

File: alvr/client_opengl/cpp/staging_pass.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    void Render() const;
+
+    gl_render_utils::Texture *GetOutputTexture() { return mExpandedTexture.get(); }
+
+private:
+
+    gl_render_utils::Texture *mInputSurface;
+    std::unique_ptr<gl_render_utils::Texture> mExpandedTexture;
+    std::unique_ptr<gl_render_utils::RenderState> mExpandedTextureState;
```
Comment (zarik5) [inline alvr/client_opengl/cpp/staging_pass.h:22]: Rename to `mOutputTexture[State]`

File: alvr/client_openxr/src/lib.rs [hunk 3]
```diff
@@ -361,6 +361,25 @@ pub fn entry_point() {
...
+            position: xr::Vector3f::default(),
+        },
+        fov: xr::Fovf {
+            angle_left: -0.1,
+            angle_right: 0.1,
+            angle_up: 0.1,
+            angle_down: -0.1,
+        },
+    };
+    let mut last_good_views = vec![ default_view, default_view ];
```
Comment (zarik5) [inline alvr/client_openxr/src/lib.rs:381]: This line doesn't look formatted correctly with `cargo fmt` but CI doesn't complain?

File: alvr/client_opengl/cpp/graphics.cpp [hunk 4]
```diff
@@ -660,9 +670,13 @@ void renderEye(
...
-void ovrRenderer_RenderFrame(ovrRenderer *renderer, const EyeInput input[2], bool isLobby) {
-    if (renderer->enableFFR) {
-        renderer->ffr->Render();
+void ovrRenderer_RenderFrame(ovrRenderer *renderer, const EyeInput input[2], bool isLobby, bool isReprojected) {
+    if (!isLobby && !isReprojected) {
+        if (renderer->enableFFR) {
+          renderer->ffr->Render();
+        } else {
+          renderer->staging_pass->Render();
+        }
```
Comment (zarik5) [inline alvr/client_opengl/cpp/graphics.cpp:679]: Always run `renderer->staging_pass->Render();` first if `!isLobby` and pipe the output to FFR. You can modify ffr.cpp so it never has to deal with OES textures. You could also run the staging pass even when `isLobby`. This should simplify the code a bit.

File: alvr/client_opengl/cpp/graphics.cpp [hunk 5]
```diff
@@ -820,22 +836,28 @@ void renderLobbyNative(const EyeInput eyeInputs[2]) {
...
+
+    EyeInput eyeInputs[2] = {};
+    eyeInputs[0].swapchainIndex = swapchainIndices[0];
+    eyeInputs[1].swapchainIndex = swapchainIndices[1];
+
+    if (streamHardwareBuffer == 0) {
+      ovrRenderer_RenderFrame(g_ctx.streamRenderer.get(), eyeInputs, false, true);
+      return;
+    }
+
```
Comment (zarik5) [inline alvr/client_opengl/cpp/graphics.cpp:852]: I would make only one call to `ovrRenderer_RenderFrame`, with the edits explained above.

File: alvr/client_openxr/src/lib.rs (hunk 1 above)
Comment (zarik5) [reply alvr/client_openxr/src/lib.rs:645]: I think this is good, no worries.

File: alvr/client_openxr/src/lib.rs [hunk 6]
```diff
@@ -603,27 +622,14 @@ pub fn entry_point() {
...
-                    vec![default_view, default_view]
-                });
+                views = if hardware_buffer.is_null() || history_views.is_none() {
+                    vec![last_good_views[0], last_good_views[1]]
+                } else {
+                    let unwrapped_views = history_views.unwrap();
+                    last_good_views[0] = unwrapped_views[0];
+                    last_good_views[1] = unwrapped_views[1];
+                    unwrapped_views
+                };
```
Comment (zarik5) [inline alvr/client_openxr/src/lib.rs:632]: I would rewrite this as:
```suggestion
                views = if let Some(views) = history_views {
                    last_good_views = views.clone();
                    views
                } else {
                    last_good_views.clone();
                };
```
Avoid using `.unwrap()` like a plague :)

**Review by zarik5 — CHANGES_REQUESTED** [review]

**Comment by galister** [comment]
i think that covers everything? ended up being able to simplify it well without getting the lobby involved.

**Comment by zarik5** [comment]
Looks good! I won't nitpick any more, it can be shipped for now :)

**Review by zarik5 — APPROVED** [review]

**Pull request merged**